cmd/az-log/cmd/common.go

```diff
@@ -0,0 +1,241 @@
+
+	v1PodLogOptions := v1.PodLogOptions{
+		Container: container,
+		Follow:    isFollow,
+	}
+
+	// If since/sinceTime is specified, data type conversion and set up PodLogOptions
+	if since != "" {
+		d, err := time.ParseDuration(since)
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(0)
+		}
+		timeDuration := int64(d.Seconds())
+		v1PodLogOptions.SinceSeconds = &timeDuration
+	} else if sinceTime != "" {
+		t, err := TimestampFormatValidation(sinceTime)
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(0)
+		}
+		timestamp := metav1.NewTime(t)
+		v1PodLogOptions.SinceTime = &timestamp
+	}
+
+	podLogOptions := make([]v1.PodLogOptions, 0)
+
+	// If logs from previous container is needed
+	if isPrevious {
+		v1PodLogOptions.Previous = true
+		podLogOptions = append(podLogOptions, v1PodLogOptions)
+		v1PodLogOptions.Previous = false
+	}
+
+	podLogOptions = append(podLogOptions, v1PodLogOptions)
+
+	for i := 0; i < len(podLogOptions); i++ {
+		req := clientsetK8s.CoreV1().Pods(consts.ReleaseNamespace).GetLogs(podName, &podLogOptions[i])
+		podLogs, err := req.Stream(context.TODO())
+		if err != nil {
+			panic(err.Error())
+		}
+
+		defer podLogs.Close()
+
+		buf := bufio.NewScanner(podLogs)
+		LogFilter(buf, volumes, nodes, requestIds, "")
+	}
+}
+
+func TimestampFormatValidation(sinceTime string) (time.Time, error) {
+	var t time.Time
+	var err error
+
+	// sinceTime input validation and convert it to time.Time from string
+	if isMatch, _ := regexp.MatchString(RFC3339Format, sinceTime); isMatch {
+		t, err = time.Parse(time.RFC3339, sinceTime)
+		if err != nil {
+			return t, fmt.Errorf("error: %v", err)
+		}
+	} else if isMatch, _ := regexp.MatchString(KlogTimeFormat, sinceTime); isMatch {
+		if isUTC, _ := regexp.MatchString(KlogTimeFormat+`$`, sinceTime); isUTC {
+			sinceTime += "Z"
+		}
+
+		t, err = time.Parse("20060102 15:04:05Z07:00", fmt.Sprint(time.Now().Year())+sinceTime)
+		if err != nil {
+			return t, fmt.Errorf("error: %v", err)
+		}
+
+	} else {
+		return t, fmt.Errorf("\"%v\" is not a valid timestamp format", sinceTime)
+	}
+
+	return t, nil
+}
+
+func LogTimeFilter(log string, sinceTime string) bool {
+	isMatch, _ := regexp.MatchString(KlogTimeFormat, log)
+	return isMatch && len(log) > 21 && log[1:21] >= sinceTime
+}
+
+func LogFilter(buf *bufio.Scanner, volumes []string, nodes []string, requestIds []string, sinceTime string) {
+	var isAfterTime bool
+	if sinceTime == "" {
+		isAfterTime = true
+	} else {
+		isAfterTime = false
+	}
+
+	for buf.Scan() {
+		log := buf.Text()
+
+		if !isAfterTime {
+			isAfterTime = LogTimeFilter(log, sinceTime)
+		}
+
+		if isAfterTime {
+			isPrint := true
+			if len(volumes) > 0 {
+				isPrint = false
+				for _, v := range volumes {
+					isPrint = strings.Contains(log, v)
+					if isPrint {
+						break
+					}
+				}
+			}
+
+			if !isPrint {
+				fmt.Println("No logs are queried")
```

Comment:
The error message could be more descriptive, maybe something like: "No logs found for volumes: <insert-volume-name> "

Reply:
I found printing this message is an error (as well as os.Exit(0)). It should just mean that log isn't not queried rather than all logs. I fixed the code.